pilot/pkg/model/push_context.go

```diff
@@ -1608,30 +1608,30 @@ func (ps *PushContext) mergeGateways(proxy *Proxy) *MergedGateway {
 
 // pre computes gateways for each network
 func (ps *PushContext) initMeshNetworks() {
-	if ps.Networks == nil || len(ps.Networks.Networks) == 0 {
-		return
-	}
-
 	ps.networkGateways = map[string][]*Gateway{}
-	for network, networkConf := range ps.Networks.Networks {
-		gws := networkConf.Gateways
-		if len(gws) == 0 {
-			// all endpoints in this network are reachable directly from others. nothing to do.
-			continue
-		}
 
-		registryNames := getNetworkRegistries(networkConf)
-		gateways := make([]*Gateway, 0, len(gws))
+	// First, if a gateway address is provided in the configuration use it. If the gateway address
+	// in the config was a hostname it got already resolved and replaced with an IP address
+	// when loading the config
+	if ps.MeshNetworks != nil {
+		for network, networkConf := range ps.MeshNetworks.Networks {
+			gws := networkConf.Gateways
+			for _, gw := range gws {
+				if gwIP := net.ParseIP(gw.GetAddress()); gwIP != nil {
+					ps.networkGateways[network] = append(ps.networkGateways[network], &Gateway{gw.GetAddress(), gw.Port})
+				}
+			}
 
-		for _, gw := range gws {
-			gateways = append(gateways, getGatewayAddresses(gw, registryNames, ps.ServiceDiscovery)...)
 		}
+	}
 
-		log.Debugf("Endpoints from registries %v on network %v reachable through %d gateways",
-			registryNames, network, len(gateways))
-
-		ps.networkGateways[network] = gateways
+	// Second, try to find gateway addresses by the provided service name
+	for network, gateways := range ps.ServiceDiscovery.NetworkGateways() {
+		// - the internal map of label gateways - these get deleted if the service is deleted, updated if the ip changes etc.
+		// - the computed map from meshNetworks (triggered by reloadNetworkLookup, the ported logic from getGatewayAddresses)
+		ps.networkGateways[network] = append(ps.networkGateways[network], gateways...)
```

Comment:
Do we want this to override `MeshNetworks`? I thought the behavior would be the opposite ... `MeshNetworks` overrides, no?

Reply:
If we consider MeshNetworks deprecated it would make sense for the "new" way to override right?

Even if not - generally seems like annotations take precedence in our APIs, although this one does feel slightly different

Reply:
Labels are applied at a more fine-grained level. I figured that's what would take precedence compared to a global setting.

Reply:
Yep I'm fine with that ... figured I'd just ask the question.